File: npbloom-web/src/__tests__/ui/content/editNodes.test.ts
import { describe, expect, it } from 'vitest';
import { Sentence } from '../../../types';
import {
  idMap, NodeIndicatorInPlot, PlotCoordsOffset, set, StringSlice, TreeCoordsOffset, UnpositionedBranchingNode,
  UnpositionedTerminalNode, UnpositionedTree, jsNodeMapRepr
} from 'npbloom-core';
import { handleLocalSentenceChange, newNodeFromSelection } from '../../../ui/content/editNodes';

describe('high-level tree editing routines', () => {
  describe('defining new nodes', () => {
    it('defines a new branching node', () => {
      expect(newNodeFromSelection({
        nodeIndicators: [new NodeIndicatorInPlot('tree', 'a'), new NodeIndicatorInPlot('tree', 'b')]
      }, '')).toMatchObject({ targetChildIds: set(['a', 'b']) });
    });

    describe('defines a new terminal node', () => {
      const testSentencesForWordRanges: [Sentence, number, StringSlice][] = [
        // The word in...                          ...at position...     ...is in range...
        ['She will buy the red sweater.',          2,                    new StringSlice(0, 3)],
        ['I have nowhere to go.',                  18,                   new StringSlice(18, 20)],
        ['That was a big surprise, wasn\'t it?',   27,                   new StringSlice(25, 31)],
        ['Look at the blackboard!',                13,                   new StringSlice(12, 22)],
        ['Look\u2003at\u2003the\u2003blackboard!', 13,                   new StringSlice(12, 22)],
      ];

      it.each(testSentencesForWordRanges)(
        'finds that the word in "%s" intersecting position %d is in range %p',
        (sentence, position, expectedSlice) => {
          expect(newNodeFromSelection({ treeId: 'test', slice: new StringSlice(position, position) }, sentence))
            .toMatchObject({ targetSlice: expectedSlice, triangle: false });
        });
    });
  });

  describe('accommodating sentence changes', () => {
    const tree = new UnpositionedTree(
      'The dog jumped.',
      idMap({
        'branch1': new UnpositionedBranchingNode('NP', new TreeCoordsOffset(0, 0), set(['term1', 'term2'])),
        'term1': new UnpositionedTerminalNode('Det', new TreeCoordsOffset(0, 0), new StringSlice(0, 3)),
        'term2': new UnpositionedTerminalNode('N', new TreeCoordsOffset(0, 0), new StringSlice(4, 7)),
        'term3': new UnpositionedTerminalNode('VP', new TreeCoordsOffset(0, 0), new StringSlice(8, 14), true),
        'top': new UnpositionedBranchingNode('S', new TreeCoordsOffset(0, 5), set(['branch1', 'term3'])),
      }),
      new PlotCoordsOffset(0, 0),
    );

    it('sets the sentence associated with a tree', () => {
      expect(handleLocalSentenceChange('The dogs jumped.', new StringSlice(7, 7))(tree).sentence).toBe('The dogs jumped.');
    });

    it.each([
      ['The dogs jumped.', new StringSlice(7, 7)],
      ['The wdog jumped.', new StringSlice(4, 4)],
      ['The doug jumped.', new StringSlice(6, 6)],
      ['The og jumped.', new StringSlice(5, 5)],  // Backward delete from index 5
      ['The og jumped.', new StringSlice(4, 4)],  // Forward delete from index 4
      ['The dg jumped.', new StringSlice(6, 6)],  // Backward delete from index 6
      ['The dg jumped.', new StringSlice(5, 5)],  // Forward delete from index 5
      ['The do jumped.', new StringSlice(7, 7)],  // Backward delete from index 7
      ['The do jumped.', new StringSlice(6, 6)],  // Forward delete from index 6
      ['The g jumped.', new StringSlice(6, 6)],  // Backward word delete from index 6
      ['The d jumped.', new StringSlice(5, 5)],  // Forward word delete from index 5
    ] as [Sentence, StringSlice][])('reassigns terminal node slices when changing sentence to "%s"', (newSentence, oldSelection) => {
      expect(jsNodeMapRepr(handleLocalSentenceChange(newSentence, oldSelection)(tree).nodes)).toMatchSnapshot();
    });
  });
});

File: npbloom-core/src/commonTest/kotlin/ui/content/EditNodesTest.kt
package ui.content

import content.EntitySet
import content.StringSlice
import content.unpositioned.*
import ui.SliceSelectionInPlot
import kotlin.test.Test
import kotlin.test.assertEquals

class EditNodesTest {
    private val tree = UnpositionedTree(
        id = "pq6dLbe",
        "The dog jumped.",
        EntitySet(
            UnpositionedBranchingNode("branch1", "NP", TreeCoordsOffset(0.0, 0.0), setOf("term1", "term2")),
            UnpositionedTerminalNode("term1", "Det", TreeCoordsOffset(0.0, 0.0), StringSlice(0, 3)),
            UnpositionedTerminalNode("term2", "N", TreeCoordsOffset(0.0, 0.0), StringSlice(4, 7)),
            UnpositionedTerminalNode("term3", "VP", TreeCoordsOffset(0.0, 0.0), StringSlice(8, 14), true),
            UnpositionedBranchingNode("top", "S", TreeCoordsOffset(0.0, 5.0), setOf("branch1", "term3")),
        ),
        PlotCoordsOffset(0.0, 0.0),
    )

    @Test
    fun newNodeFromSelectedWord() =
        assertEquals(
            InsertedTerminalNode("UE7q2t37", "", null, StringSlice(0, 4), triangle = false),
            newNodeFromSelection("UE7q2t37", SliceSelectionInPlot("YC38BV4q", StringSlice(0, 4)), "Alex baked cookies.")
        )

    @Test
    fun newNodeFromSelectedPhrase() =
        assertEquals(
            InsertedTerminalNode("I1cw913", "", null, StringSlice(5, 18), triangle = true),
            newNodeFromSelection("I1cw913", SliceSelectionInPlot("YC38BV4q", StringSlice(5, 18)), "Alex baked cookies.")
        )

    @Test
    fun newNodeFromZeroLengthSlice() =
        assertEquals(
            InsertedTerminalNode("v1M9Wei9", "", null, StringSlice(0, 4), triangle = false),
            newNodeFromSelection("v1M9Wei9", SliceSelectionInPlot("YC38BV4q", StringSlice(3, 3)), "Alex baked cookies.")
        )

    @Test
    fun handleLocalSentenceChange() =
        listOf(
            Triple("The dogs jumped.", StringSlice(7, 7), StringSlice(4, 7)),
            Triple("The wdog jumped.", StringSlice(4, 4), StringSlice(5, 8)),
            Triple("The doug jumped.", StringSlice(6, 6), StringSlice(4, 8)),
            Triple("The blue dog jumped.", StringSlice(4, 4), StringSlice(9, 12)),
            Triple("The dog quickly jumped.", StringSlice(8, 8), StringSlice(4, 7)),
            Triple("The og jumped.", StringSlice(5, 5), StringSlice(4, 6)),  // Backward delete from index 5
            Triple("The og jumped.", StringSlice(4, 4), StringSlice(4, 6)),  // Forward delete from index 4
            Triple("The dg jumped.", StringSlice(6, 6), StringSlice(4, 6)),  // Backward delete from index 6
            Triple("The dg jumped.", StringSlice(5, 5), StringSlice(4, 6)),  // Forward delete from index 5
            Triple("The do jumped.", StringSlice(7, 7), StringSlice(4, 6)),  // Backward delete from index 7
            Triple("The do jumped.", StringSlice(6, 6), StringSlice(4, 6)),  // Forward delete from index 6
            Triple("The g jumped.", StringSlice(6, 6), StringSlice(4, 5)),  // Backward word delete from index 6
            Triple("The d jumped.", StringSlice(5, 5), StringSlice(4, 5)),  // Forward word delete from index 5
            Triple("The  jumped.", StringSlice(4, 7), StringSlice(4, 4)),  // Selection delete encompassing entire word
            Triple("The jumped.", StringSlice(4, 8), StringSlice(4, 4)),  // Selection delete encompassing entire word and space after it
            Triple("dog jumped.", StringSlice(0, 4), StringSlice(0, 3)),  // Selection delete before slice
            Triple("The dog", StringSlice(8, 15), StringSlice(4, 7)),  // Selection delete after slice
        ).map { (newSentence, oldSelection, newSlice) ->
            val newNode = tree.handleLocalSentenceChange(newSentence, oldSelection).node("term2")
            if (newNode is UnpositionedTerminalNode) assertEquals(newSlice, newNode.slice)
            else if (newNode is UnpositionedFormerlyTerminalNode) assertEquals(newSlice, newNode.formerSlice)
        }
}
